### ISSUE TITLE
Qt moc fails if sources have same name

### ISSUE THREAD

**Issue by ePirat** [mesonbuild/meson#3797]
Similar to #3791, the Qt modules fails with moc files that have the same name:

```
moc_sources = files('main_interface.hpp',
    'dialogs/playlist.cpp',
    'components/playlist/playlist.cpp')

qt5pre_files = qt5.preprocess(moc_sources : moc_sources)
```

```
ERROR: Multiple producers for Ninja target "modules/modules@@qt_plugin@sha/playlist.moc". Please rename your targets.
```

**Comment by jpakkane** [comment]
The reason this happens is that the generator output name is `@BASENAME@` or something similar. If you have two files with identical filenames their `@BASENAME@`s are the same and thus will collide.

I can't think of what the true and correct solution for this would be. In general it is considered poor form to have many files in your project with the exact same name. It is very confusing because sooner or later you will mix the two up. Obviously renaming files is not a simple solution for you.

A workaround is to compile files with identical names (in this case e.g. `dialogs/playlist.cpp`) in a helper static library and link that to your final target. It's a bit messy but should at least unblock your work for this.

**Comment by jpakkane** [comment]
That's because those use a generator and that uses `@BASENAME@` which is defined to not have those path segments and people actually depend on that (I think).

**Comment by equeim** [comment]
Yeah, it will break moc_source and ui_files arguments. Maybe instead of appending directory prefixes to object files Meson should create an actual directory structure (setting correct includes to current directory), like Waf does. This will solve the issue with same file names for generators too.

**Comment by PPPPoint** [comment]
Does this promblem solve？

**Comment by bonzini** [comment]
This can now be done with `preserve_paths` (f4f50db44155a4e050c58a74849ba8859e5daf51).
